operators/src/pro/cache/tile_cache.rs

```diff
@@ -29,16 +30,425 @@ pub struct TileCacheBackend {
+        ensure!(
+            self.byte_size_used >= bytes,
+            crate::pro::cache::error::NegativeSizeOfCache
+        );
+        self.remove_bytes(bytes);
+        Ok(())
+    }
+
+    #[inline]
+    pub fn try_remove_element_bytes<T: ByteSize>(&mut self, element: &T) -> Result<(), CacheError> {
+        let bytes = element.byte_size();
+        self.try_remove_bytes(bytes)
+    }
+
+    #[inline]
+    fn add_bytes_allow_overflow(&mut self, bytes: usize) {
+        self.byte_size_used += bytes;
+        if self.byte_size_used > self.byte_size_total {
+            trace!(
+                "overflowing cache size by {} bytes, total size: {}, added bytes: {}",
+                self.byte_size_used - self.byte_size_total,
+                self.byte_size_total,
+                bytes
+            );
+        }
+    }
+}
+
+type CacheSizeLruRef<'a> = (
+    &'a mut HashMap<CacheEntryId, CacheEntry>,
+    &'a mut TileCacheBackendSize,
+    &'a mut LruCache<CacheEntryId, CanonicOperatorName>,
+);
+
+type LandingZoneSizeRef<'a> = (
+    &'a mut HashMap<QueryId, LandingZoneEntry>,
+    &'a mut TileCacheBackendSize,
+);
+
+impl TileCacheBackend {
+    fn cache_and_size_and_lru_mut(&mut self, key: &CanonicOperatorName) -> Option<CacheSizeLruRef> {
+        self.operator_caches
+            .get_mut(key)
+            .map(|cache| (&mut cache.entries, &mut self.cache_size, &mut self.lru))
+    }
+
+    fn landing_zone_and_size_mut(
+        &mut self,
+        key: &CanonicOperatorName,
+    ) -> Option<LandingZoneSizeRef> {
+        self.operator_caches
+            .get_mut(key)
+            .map(|cache| (&mut cache.landing_zone, &mut self.landing_zone_size))
+    }
+
+    pub fn query_cache<T>(
+        &mut self,
+        key: &CanonicOperatorName,
+        query: &RasterQueryRectangle,
+    ) -> Option<CacheTileStream<T>>
+    where
+        T: Pixel + Cachable,
+    {
+        let (cache, _cache_size, cache_lru) = self.cache_and_size_and_lru_mut(key)?;
+
+        let mut expired_ids = Vec::new();
+        let x = cache.iter().find(|&(id, r)| {
+            if r.is_expired() {
+                expired_ids.push(*id);
+                return false;
+            }
+            r.matches(query)
+        });
+
+        let res = if let Some((id, entry)) = x {
+            // set as most recently used
+            cache_lru.promote(id);
+            T::stream(entry.tiles.tile_stream(query))
+        } else {
+            None
+        };
+
+        // remove expired entries
+        self.remove_querys_from_cache_and_lru(key, expired_ids.as_slice());
+
+        res
+    }
+
+    pub fn insert_query<T: Pixel + Cachable>(
+        &mut self,
+        key: &CanonicOperatorName,
+        query: &RasterQueryRectangle,
+    ) -> Result<QueryId, CacheError> {
+        // if there is no cache for this operator graph, we create one. However, it is not clear yet, if we will actually cache anything.
```

Comment:
why not? You mean because the query insertion may be aborted?

Reply:
i think this is a comment from your code